File: app/src/main/java/com/example/collections_and_maps/models/benchmarks/CollectionsBenchmark.java
package com.example.collections_and_maps.models.benchmarks;

import static com.example.collections_and_maps.models.benchmarks.ResultItem.EMPTY;

import android.util.Log;

import com.example.collections_and_maps.R;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.CopyOnWriteArrayList;

public class CollectionsBenchmark implements Benchmark {

    private final Random random = new Random();

    private final int[] listNamesForHead = new int[]{R.string.ArrayList, R.string.LinkedList, R.string.CopyOnWrite};
    private final int[] listNamesForMethod = new int[]{R.string.add_begin, R.string.add_middle,
            R.string.add_end, R.string.search_value, R.string.remove_begin,
            R.string.remove_middle, R.string.remove_end};

    @Override
    public List<ResultItem> getItemsList(boolean showProgress) {
        List<ResultItem> itemsList = new ArrayList<>();

        for (int itemOfListHead : listNamesForHead) {
            itemsList.add(new ResultItem(itemOfListHead, R.string.empty, EMPTY, false));
        }

        for (int methodsID : listNamesForMethod) {
            itemsList.add(new ResultItem(R.string.empty, methodsID, EMPTY, false));
            for (int headsID : listNamesForHead) {
                itemsList.add(new ResultItem(headsID, methodsID, EMPTY, showProgress));
            }
        }
        return itemsList;
    }


    @Override
    public double getMeasureTime(ResultItem rItem, int value) {
        final List<Integer> list;
        switch (rItem.headerText) {
            case R.string.ArrayList:
                list = new ArrayList<>(Collections.nCopies(value, 0));
                break;
            case R.string.LinkedList:
                list = new LinkedList<>(Collections.nCopies(value, 0));
                break;
            case R.string.CopyOnWrite:
                list = new CopyOnWriteArrayList<>(Collections.nCopies(value, 0));
                break;
            default:
                throw new IllegalStateException("Unexpected value: " + rItem.headerText);
        }


        return calculateResult(rItem.methodName, list);
    }


    private double calculateResult(int methodName, List<Integer> list) {
        switch (methodName) {
            case R.string.add_begin:
                return addItemToStart(list);
            case R.string.add_middle:
                return addItemToMiddle(list);
            case R.string.add_end:
                return addItemToEnd(list);
            case R.string.search_value:
                return searchByValue(list);
            case R.string.remove_begin:
                return removingInBeginning(list);
            case R.string.remove_middle:
                return removingInMiddle(list);
            case R.string.remove_end:
                return removingInEnd(list);
            default:
                throw new IllegalArgumentException();
        }
    }


    private double addItemToStart(List<Integer> list) {
        double start = System.nanoTime();
        list.add(0);
        double resultTime = System.nanoTime() - start;
        return resultTime;
    }

    private double addItemToMiddle(List<Integer> list) {
        double start = System.nanoTime();
        list.add(list.size() / 2, 0);
        double resultTime = System.nanoTime() - start;
//        Log.i(TAG, "addItemToMiddle = " + resultTime);
        return resultTime;
    }

    private double addItemToEnd(List<Integer> list) {
        double start = System.nanoTime();
        list.add(list.size(), 0);
        double resultTime = System.nanoTime() - start;
//        Log.i(TAG, "addItemToEnd = " + resultTime);
        return resultTime;
    }

    private double searchByValue(List<Integer> list) {
        int index = random.nextInt(list.size());
        list.add(index, index);
        double start = System.nanoTime();
        boolean has = list.contains(index);
        double resultTime = System.nanoTime() - start;
//        Log.i(TAG, "searchByValue = " + resultTime);
        return resultTime;
    }

    private double removingInBeginning(List<Integer> list) {
        double start = System.nanoTime();
        list.remove(0);
        double resultTime = System.nanoTime() - start;
//        Log.i(TAG, "removingInBeginning = " + resultTime);
        return resultTime;
    }

    private double removingInMiddle(List<Integer> list) {
        double start = System.nanoTime();
        list.remove(list.size() / 2);
        double resultTime = System.nanoTime() - start;
//        Log.i(TAG, "removingInMiddle = " + resultTime);
        return resultTime;
    }

    private double removingInEnd(List<Integer> list) {
        double start = System.nanoTime();
        list.remove(list.size() - 1);
        double resultTime = System.nanoTime() - start;
//        Log.i(TAG, "removingInEnd = " + resultTime);
        return resultTime;
    }

    @Override
    public int getSpan() {
        return listNamesForHead.length;
    }
}

File: app/src/main/java/com/example/collections_and_maps/models/benchmarks/CollectionsBenchmark.kt
package com.example.collections_and_maps.models.benchmarks

import com.example.collections_and_maps.R
import java.util.Collections
import java.util.LinkedList
import java.util.Random
import java.util.concurrent.CopyOnWriteArrayList

open class CollectionsBenchmark : Benchmark {

    private val random = Random()
    private val listNamesForHead =
        intArrayOf(R.string.ArrayList, R.string.LinkedList, R.string.CopyOnWrite)
    private val listNamesForMethod = intArrayOf(
        R.string.add_begin, R.string.add_middle,
        R.string.add_end, R.string.search_value, R.string.remove_begin,
        R.string.remove_middle, R.string.remove_end
    )

    override fun getItemsList(showProgress: Boolean): List<ResultItem> {
        val itemsList: MutableList<ResultItem> = ArrayList()
        for (head in listNamesForHead) {
            itemsList.add(ResultItem(head, R.string.empty, ResultItem.EMPTY, false))
        }
        for (methodsID in listNamesForMethod) {
            itemsList.add(ResultItem(R.string.empty, methodsID, ResultItem.EMPTY, false))
            for (headsID in listNamesForHead) {
                itemsList.add(ResultItem(headsID, methodsID, ResultItem.EMPTY, showProgress))
            }
        }
        return itemsList
    }

    override fun getMeasureTime(rItem: ResultItem, value: Int): Double {
        val list: MutableList<Int> = when (rItem.headerText) {
            R.string.ArrayList -> ArrayList(Collections.nCopies(value, 0))
            R.string.LinkedList -> LinkedList(Collections.nCopies(value, 0))
            R.string.CopyOnWrite -> CopyOnWriteArrayList(Collections.nCopies(value, 0))
            else -> throw IllegalStateException("Unexpected value: " + rItem.headerText)
        }
        return calculateResult(rItem.methodName, list)
    }

    override fun getSpan(): Int = listNamesForHead.size



    private fun calculateResult(methodName: Int, list: MutableList<Int>): Double {
        return when (methodName) {
            R.string.add_begin -> addItemToStart(list)
            R.string.add_middle -> addItemToMiddle(list)
            R.string.add_end -> addItemToEnd(list)
            R.string.search_value -> searchByValue(list)
            R.string.remove_begin -> removingInBeginning(list)
            R.string.remove_middle -> removingInMiddle(list)
            R.string.remove_end -> removingInEnd(list)
            else -> throw IllegalArgumentException()
        }
    }

    private fun addItemToStart(list: MutableList<Int>): Double {
        val start = System.nanoTime().toDouble()
        list.add(0)
        return System.nanoTime() - start
    }

    private fun addItemToMiddle(list: MutableList<Int>): Double {
        val start = System.nanoTime().toDouble()
        list.add(list.size / 2, 0)
        return System.nanoTime() - start
    }

    private fun addItemToEnd(list: MutableList<Int>): Double {
        val start = System.nanoTime().toDouble()
        list.add(list.size, 0)
        return System.nanoTime() - start
    }

    private fun searchByValue(list: MutableList<Int>): Double {
        val index = random.nextInt(list.size)
        list.add(index, index)
        val start = System.nanoTime().toDouble()
        val has = list.contains(index)
        return System.nanoTime() - start
    }

    private fun removingInBeginning(list: MutableList<Int>): Double {
        val start = System.nanoTime().toDouble()
        list.removeAt(0)
        return System.nanoTime() - start
    }

    private fun removingInMiddle(list: MutableList<Int>): Double {
        val start = System.nanoTime().toDouble()
        list.removeAt(list.size / 2)
        return System.nanoTime() - start
    }

    private fun removingInEnd(list: MutableList<Int>): Double {
        val start = System.nanoTime().toDouble()
        list.removeAt(list.size - 1)
        return System.nanoTime() - start
    }
}
